packages/astro/src/vite-plugin-astro/index.ts

```diff
@@ -27,85 +20,28 @@ interface AstroPluginOptions {
-
-	function relativeToRoot(pathname: string) {
-		const arg = startsWithForwardSlash(pathname) ? '.' + pathname : pathname;
-		const url = new URL(arg, config.root);
-		return slash(fileURLToPath(url)) + url.search;
-	}
 
-	function resolveRelativeFromAstroParent(id: string, parsedFrom: ParsedRequestResult): string {
-		const filename = normalizeFilename(parsedFrom.filename, config);
-		const resolvedURL = new URL(id, `file://${filename}`);
-		const resolved = resolvedURL.pathname;
-		if (isBrowserPath(resolved)) {
-			return relativeToRoot(resolved + resolvedURL.search);
-		}
-		return slash(fileURLToPath(resolvedURL)) + resolvedURL.search;
-	}
-
-	return {
+	const prePlugin: vite.Plugin = {
 		name: 'astro:build',
 		enforce: 'pre', // run transforms before other plugins can
 		configResolved(_resolvedConfig) {
 			resolvedConfig = _resolvedConfig;
 		},
-		// note: don’t claim .astro files with resolveId() — it prevents Vite from transpiling the final JS (import.meta.glob, etc.)
-		async resolveId(id, from, opts) {
-			// If resolving from an astro subresource such as a hoisted script,
-			// we need to resolve relative paths ourselves.
-			if (from) {
-				const parsedFrom = parseAstroRequest(from);
-				const isAstroScript = parsedFrom.query.astro && parsedFrom.query.type === 'script';
-				if (isAstroScript && isRelativePath(id)) {
-					return this.resolve(resolveRelativeFromAstroParent(id, parsedFrom), from, {
-						custom: opts.custom,
-						skipSelf: true,
-					});
-				}
-			}
-
-			// serve sub-part requests (*?astro) as virtual modules
-			const { query } = parseAstroRequest(id);
-			if (query.astro) {
-				// TODO: Try to remove these custom resolve so HMR is more predictable.
-				// Convert /src/pages/index.astro?astro&type=style to /Users/name/
-				// Because this needs to be the id for the Vite CSS plugin to property resolve
-				// relative @imports.
-				if (query.type === 'style' && isBrowserPath(id)) {
-					return relativeToRoot(id);
-				}
-				// Strip `/@fs` from linked dependencies outside of root so we can normalize
-				// it in the condition below. This ensures that the style module shared the same is
-				// part of the same "file" as the main Astro module in the module graph.
-				// "file" refers to `moduleGraph.fileToModulesMap`.
-				if (query.type === 'style' && id.startsWith('/@fs')) {
-					id = removeLeadingForwardSlashWindows(id.slice(4));
-				}
-				// Convert file paths to ViteID, meaning on Windows it omits the leading slash
-				if (isFullFilePath(id)) {
-					return viteID(new URL('file://' + id));
-				}
-				return id;
-			}
-		},
 		async load(id, opts) {
 			const parsedId = parseAstroRequest(id);
 			const query = parsedId.query;
 			if (!query.astro) {
 				return null;
 			}
 			// For CSS / hoisted scripts, the main Astro module should already be cached
-			const filename = normalizeFilename(parsedId.filename, config);
```

Comment:
This still needs a `normalizeFilename` in `load` as it may receive a root-relative path, like `/src/components/Foo.astro`. This seems to only happen in tests (likely the mocked fs used?), but I leave it here for now just in case.